### PR TITLE
Add icu to macOS, fix Big Sur (11.x) dependencies

### PR DESCRIPTION
**Pull request by adam-bloom**
FYI @LRFLEW

I'll be opening up an OpenRCT2 PR soon to use these updated dependencies in the macOS cmake builds. This will let us run the tests in macOS CI. (It will be added as an additional CI run to the existing Xcode run for now.)

In order to complete that, I'll need a new release with these changes.

### CODE REVIEW

File: openrct2-ports/icu/portfile.cmake
Comment (adam-bloom) [inline openrct2-ports/icu/portfile.cmake:202]: This is the block of changes from the main icu port. We could also do this as a patch file, similar to https://github.com/OpenRCT2/Dependencies/blob/master/0001-Patch-duktape.patch, and add the patch application to the build. Let me know if that would be preferred and I'll switch this over to go that route instead of overlay-ports.

Comment (LRFLEW) [reply openrct2-ports/icu/portfile.cmake:202]: Is this a change that can be submitted to the upstream vcpkg project. I'd be hesitant to use a patch or custom port for this sort of thing as vcpkg changes often, and other users could benefit from our fixes. I usually find that PRs can be merged within about a week if you keep up with requests they make.

Comment (adam-bloom) [reply openrct2-ports/icu/portfile.cmake:202]: I only added the fixes for the two icu libraries we use - the upstream port would need to patch all of them. I wasn't entirely sure if this was even the best way to patch them, or if the icu configure scripts for --enable-rpath should be enhanced. I'm still learning the ins and outs of rpath, so I figured it'd be better to patch it ourselves, figure out what to do upstream, and if upstream merges a change, we can remove our patch.

As you can probably tell, I modeled this after your openssl port changes, so perhaps changing this after the build like this is ok? I'll just need to add a loop and cover all of the icu modules.

Comment (adam-bloom) [reply openrct2-ports/icu/portfile.cmake:202]: This looks to be the issue upstream, btw:
https://github.com/unicode-org/icu/blob/master/icu4c/source/config/mh-darwin#L32-L37
With --enable-rpath on, the install dir is used instead of @rpath. I tried passing `-libdir=@rpath` to configure, but it expects an absolute path for libdir if it is passed.

**Comment by adam-bloom** [comment]
The icu port includes all icu libraries. We only need uc and data (uc dependency). Should we remove the others? I did not see any options in the icu makefiles to exclude building modules. Or should we just include the extra libraries?

**Comment by LRFLEW** [comment]
My opinion (so take it with a grain of salt): This repo should distribute everything that vcpkg produces in the "installed" directory, and the main project can link against just what it needs. I don't like the idea of hardcoding "build, then remove these specific files", as it's a system that could break by vcpkg changes.

Probably the best way of doing it is, if icu's build system supports independently built components, to add vcpkg port features for each disable-able build component. Since we have our own copy of the port, this is something we can do locally, but I still think we should try to get any changes we make here upstream. In my experience, the vcpkg team is pretty open to partial fixes (eg. only some of icu's component libraries being fixed) so long as it's not a *regression*.

**Review by LRFLEW — CHANGES_REQUESTED** [review]
The folder `/dummy-port` was added as part of the "use system zlib/bzip2" stuff, and can be removed in this PR as well.

File: openrct2-ports/icu/portfile.cmake
Comment (LRFLEW) [reply openrct2-ports/icu/portfile.cmake:202]: If this is an upstream issue, then a) Making a patch to the icu project might be a good idea (probably). b) This seems like a good opportunity for a port patch. In [the icu port directory](https://github.com/microsoft/vcpkg/tree/master/ports/icu) there are a bunch of .patch files that get applied to the base project in `portfile.cmake`. If it's a matter of just changing `-install_name $(libdir)/$(notdir $(MIDDLE_SO_TARGET)` to `-install_name @rpath/$(notdir $(MIDDLE_SO_TARGET)`, then that's something to consider instead of changing the install-name post-build (which isn't an optimal solution).

**Comment by adam-bloom** [comment]
> Probably the best way of doing it is, if icu's build system supports independently built components, to add vcpkg port features for each disable-able build component. 
 
Let me play around some more, but I couldn't figure this out earlier in the week. There are some flags for enabling/disabling features/packages in the configure script, but none of them seem to work for disabling modules. Let me look at their build chain in a little more detail.

File: openrct2-ports/icu/portfile.cmake
Comment (adam-bloom) [reply openrct2-ports/icu/portfile.cmake:202]: Wouldn't `VCPKG_INSTALL_NAME_DIR` be passed to icu as the `$(libdir)` already? I believe that behavior is correct already - when you pass `--enable-rpath`, the library is built with the absolute path to the library directory as its path. What we need is a new option for using the rpath prefix. I'll have to add it to the icu configure script to make it accessible at the point where we'd want to make that change. Good idea on a patch though. Let me give it a try.

The ICU project seems to have some thoughts on rpath ;) https://github.com/unicode-org/icu/blob/master/icu4c/source/configure#L6065

Comment (LRFLEW) [reply openrct2-ports/icu/portfile.cmake:202]: `VCPKG_INSTALL_NAME_DIR` is distinct from what is probably `$(libdir)`. Setting `VCPKG_INSTALL_NAME_DIR` in a triplet changes the path in the install-name to a different specified path from where it's installed to. So if it is installing `/some/install/path/libx.dylib`, but the install-name should be `/some/other/path/because/reasons/libx.dylib`, then you can add to the triplet file `set(VCPKG_INSTALL_NAME_DIR "/some/other/path/because/reasons")`. This variable basically just sets [`CMAKE_INSTALL_NAME_DIR`](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_NAME_DIR.html). The default is to use `@rpath` thanks to another CMAKE variable.

**Comment by adam-bloom** [comment]
Both speexdsp and libzip will get upstream (vcpkg) PRs and can be easily removed from openrct2-ports once those are merged. Lines added are:

libzip: https://github.com/adam-bloom/Dependencies/blob/macos_vcpkg_big_sur/openrct2-ports/libzip/portfile.cmake#L36
speexdsp: https://github.com/adam-bloom/Dependencies/blob/macos_vcpkg_big_sur/openrct2-ports/speexdsp/portfile.cmake#L44-L45
https://github.com/adam-bloom/Dependencies/blob/macos_vcpkg_big_sur/openrct2-ports/speexdsp/CMakeLists.txt#L104-L115

File: macos_build.sh
Comment (LRFLEW) [inline macos_build.sh:11]: While we're here, I set the freetype build to exclude the `brotli` feature simply to avoid freetype from triggering any extra library builds. However, since we're adding icu, zlib, and bzip2 here, there's not much reason to *not* include libbrotli, especially if it has a functional use (enabling WOFF2 fonts). This can be done by removing the entire `[...]` part for freetype and having it build all the default features.

**Comment by adam-bloom** [comment]
I opened two vcpkg PRs so we can remove the copies of `libpng` and `speexdsp`

https://github.com/microsoft/vcpkg/pull/15585
https://github.com/microsoft/vcpkg/pull/15589

If we release before those are merged, I'll open a new PR to remove the copies once those get merged upstream.

**Comment by tupaschoal** [comment]
Well, that's the two PRs on vcpkg merged :)

**Comment by adam-bloom** [comment]
Indeed! I've removed the local copies of both of those ports. For ICU, I've opened https://unicode-org.atlassian.net/browse/ICU-21458 to track the feature request. That one is likely not realistic to wait for, and I'd like to see if others in the ICU community have comments before proposing a vcpkg patch to hold us over.

**Comment by LRFLEW** [comment]
@adam-bloom You can wait for the ICU community to comment, but honestly, I'd just make the PR to vcpkg *now* so that they have a chance to comment on it, and if/when ICU comments, you can update the PR for it.

It's up to you if you want to wait, though.

**Comment by adam-bloom** [comment]
Opened https://github.com/microsoft/vcpkg/pull/15706 :)

**Comment by adam-bloom** [comment]
@IntelOrca @LRFLEW all vcpkg PRs have been merged and the custom ports have been removed. I just "released" my fork and tested the build - everything looks good. I think this is good to go.

**Comment by LRFLEW** [comment]
Took another look at it. Now that the "custom ports" has been removed, it wasn't as much to try and understand. My point about brotli still stands. Otherwise, I give this my :+1:.

**Comment by adam-bloom** [comment]
> Took another look at it. Now that the "custom ports" has been removed, it wasn't as much to try and understand. My point about brotli still stands. Otherwise, I give this my 👍.

I must have missed that comment earlier, @LRFLEW! That change has been made now.

**Review by LRFLEW — APPROVED** [review]

**Comment by adam-bloom** [comment]
@IntelOrca can we also get an official release made that includes the macOS zip file? I can repoint the OpenRCT PR to that release after that - we need to do that before it can be merged.

**Comment by IntelOrca** [comment]
@adam-bloom I have tagged, but the repo will not generate a macOS zip file. Are you able to add github actions that will generate it?

**Comment by adam-bloom** [comment]
I see @janisozaur had started a PR for that: #24. I can try to take that over if they don’t intend to finish it. I wasn’t sure if someone was in the habit of generating that macOS zip manually for now.

**Comment by IntelOrca** [comment]
> I wasn’t sure if someone was in the habit of generating that macOS zip manually for now.

I don't think we have anyone left on our team that is able to do this.

> I see @janisozaur had started a PR for that: #24. I can try to take that over if they don’t intend to finish it.

That should be fine.

**Comment by bryankeller** [comment]
@adam-bloom @LRFLEW have you (or anyone else) started to think about getting these build for the native architecture on macOS (aka `x86_64` _or_ `arm64`)?

**Comment by LRFLEW** [comment]
@bryankeller I have looked into it a bit, but I think the problems with it are upstream with the dependencies. Vcpkg can build *some things* for macOS ARM64, however it's not 100% yet. In particular, OpenSSL's build system doesn't work for macOS ARM64 builds (IIRC, it still assumes ARM = iOS). 

However, there's a more significant issue, IMO. When we add ARM64 support for macOS builds, we should *really* be doing it as a Universal 2 release. At least long-term, I don't want to split the macOS builds (in part because that can make things complicated going forward). However, vcpkg doesn't really have a way of doing that currently. CMake has a way of making that work ([CMAKE_OSX_ARCHITECTURES](https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_ARCHITECTURES.html)), but it's currently not utilized by vcpkg. I was planning on working on that myself at some point, but just haven't had much of a chance.

**Comment by bryankeller** [comment]
@LRFLEW I spent some time last night and the night before trying to get the dependencies repo building on my DTK Mac. I was going crazy trying to figure out why the builds were all resulting in x86 binaries... then I realized my terminal was running under Rosetta emulation 🤦‍♂️

So I ran the build script natively, thinking I solved the problem and would get arm64 libraries as output. That's when I ran into what you said - not all of them can build for arm64.

I agree a universal build would be ideal. Perhaps that could be configured in the OpenRCT project by giving it access to both x86 and arm64 libraries from this repo. Eventually, once the VCPKG supports building universal binaries, the OpenRCT project won't need both sets of libraries and can just reference universal ones.

**Comment by LRFLEW** [comment]
>Perhaps that could be configured in the OpenRCT project by giving it access to both x86 and arm64 libraries from this repo.

Unfortunately, that's not *quite* how this works. To link a universal build, you need universal builds of the dependencies I believe. IIRC from back when "Universal" meant PPC & Intel, there is a way of *merging* two builds, but that gets messy **fast**.

Of course, until vcpkg can build all the projects as ARM64 for macOS, this isn't something we need to figure out right away.

**Comment by IntelOrca** [comment]
@LRFLEW do you have a list of libraries that don't build for ARM?

**Pull request merged**
